hkl/diffract.py

```
@@ -112,65 +140,91 @@ def __init__(self, prefix, calc_kw=None, decision_fcn=None,
-        '''
-        logger.debug('{0.name} energy changed: {1}'.format(self, value))
+        """
+        logger.debug(f"{self.name} energy changed: {value}")
         self._calc.energy = value
         self._update_position()
 
     @property
     def calc(self):
-        '''The calculation instance'''
+        """The calculation instance"""
         return self._calc
 
     @property
     def engine(self):
-        '''The calculation engine associated with the diffractometer'''
+        """The calculation engine associated with the diffractometer"""
         return self._calc.engine
 
     # TODO so these calculations change the internal state of the hkl
-    # calculation class, which is probably not a good thing -- it becomes a
-    # problem when someone uses these functions outside of move()
+    # calculation class, which is probably not a good thing
+    # -- it becomes a problem when someone uses these functions
+    # outside of move()
 
     @pseudo_position_argument
     def forward(self, pseudo):
-        solutions = self._calc.forward_iter(start=self.position, end=pseudo,
-                                            max_iters=100)
-        logger.debug('pseudo to real: {}'.format(solutions))
+        solutions = self._calc.forward_iter(
+            start=self.position, end=pseudo, max_iters=100
+        )
+        logger.debug(f"pseudo to real: {solutions}")
         return self._decision_fcn(pseudo, solutions)
 
     @real_position_argument
     def inverse(self, real):
         self._calc.physical_positions = real
         return self.PseudoPosition(*self._calc.pseudo_positions)
 
+    def check_value(self, pos):
+        """
+        Raise exception if pos is not within limits.
+
+        In a scan, a subset of the pseudo axes may be directed,
+        which are given in a dict from a set message from the
+        bluesky RunEngine.
+
+        It is not permitted to scan both pseudo and real positioners.
```

Is this check done someplace else?

---

Yes.  Good question, it's not from the *hklpy* code, though.  Given:

```
check_limits(
            bp.scan(
                [fourc],
                fourc.tth, 10, 20,
                fourc.k, 0, 0,
                3)
        )
```

it raises from `bluesky.utils`:

```
~/Apps/anaconda/envs/hkl/lib/python3.7/site-packages/bluesky/utils.py in merge_cycler(cyc)
   1387 
   1388         if type_map['real'] and type_map['pseudo']:
-> 1389             raise ValueError("Passed in a mix of real and pseudo axis.  "
   1390                              "Can not cope, failing")
   1391         pseudo_axes = type_map['pseudo']

ValueError: Passed in a mix of real and pseudo axis.  Can not cope, failing
```

---

and tested here: https://github.com/bluesky/hklpy/blob/aaae4a5d4953676d3b1d999a84e6e89476c97e4d/hkl/tests/test_fourc.py#L136-L146

---

That is checking that we won't scan a mix (which is an internal details of scan), but presamble we could call `check_values` with a mixed dictionary?  If so then this would silently drop some of that input on the floor and say `{'k': 5, 'tth': np.inf}` is ok!

I have not waded into this code in a while so my memory is getting a bit fuzzy, sorry if I am asking daft questions.

---

Good questions!

---

pseudo positioners (that fit the engine's model, such as h, k, l for the hkl engine) should be passed as a tuple, real positioners could be checked individually - starting to get dodgy

---

You're right that we need to anticipate this method being called incorrectly, as well as from the usual callers.

---

added tests and code for this situation